zcash_primitives/src/merkle_tree.rs

```diff
@@ -0,0 +1,1071 @@
+//! Implementation of a Merkle tree of commitments used to prove the existence of notes.
+
+use byteorder::{LittleEndian, ReadBytesExt};
+use std::collections::VecDeque;
+use std::io::{self, Read, Write};
+use std::iter;
+
+use sapling::SAPLING_COMMITMENT_TREE_DEPTH;
+use serialize::{Optional, Vector};
+
+/// A hashable node within a Merkle tree.
+pub trait Hashable: Clone + Copy {
+    /// Parses a node from the given byte source.
+    fn read<R: Read>(reader: R) -> io::Result<Self>;
+
+    /// Serializes this node.
+    fn write<W: Write>(&self, writer: W) -> io::Result<()>;
+
+    /// Returns the parent node within the tree of the two given nodes.
+    fn combine(usize, &Self, &Self) -> Self;
+
+    /// Returns a blank leaf node.
+    fn blank() -> Self;
+
+    /// Returns the empty root for the given depth.
+    fn empty_root(usize) -> Self;
+}
+
+struct PathFiller<Node: Hashable> {
+    queue: VecDeque<Node>,
+}
+
+impl<Node: Hashable> PathFiller<Node> {
+    fn empty() -> Self {
+        PathFiller {
+            queue: VecDeque::new(),
+        }
+    }
+
+    fn next(&mut self, depth: usize) -> Node {
+        self.queue
+            .pop_front()
+            .unwrap_or_else(|| Node::empty_root(depth))
+    }
+}
+
+/// A Merkle tree of note commitments.
+///
+/// The depth of the Merkle tree is fixed at 32, equal to the depth of the Sapling
+/// commitment tree.
+#[derive(Clone)]
+pub struct CommitmentTree<Node: Hashable> {
+    left: Option<Node>,
+    right: Option<Node>,
+    parents: Vec<Option<Node>>,
+}
+
+impl<Node: Hashable> CommitmentTree<Node> {
+    /// Creates an empty tree.
+    pub fn new() -> Self {
+        CommitmentTree {
+            left: None,
+            right: None,
+            parents: vec![],
+        }
+    }
+
+    /// Reads a `CommitmentTree` from its serialized form.
+    pub fn read<R: Read>(mut reader: R) -> io::Result<Self> {
+        let left = Optional::read(&mut reader, |r| Node::read(r))?;
+        let right = Optional::read(&mut reader, |r| Node::read(r))?;
+        let parents = Vector::read(&mut reader, |r| Optional::read(r, |r| Node::read(r)))?;
+
+        Ok(CommitmentTree {
+            left,
+            right,
+            parents,
+        })
+    }
+
+    /// Serializes this tree as an array of bytes.
+    pub fn write<W: Write>(&self, mut writer: W) -> io::Result<()> {
+        Optional::write(&mut writer, &self.left, |w, n| n.write(w))?;
+        Optional::write(&mut writer, &self.right, |w, n| n.write(w))?;
+        Vector::write(&mut writer, &self.parents, |w, e| {
+            Optional::write(w, e, |w, n| n.write(w))
+        })
+    }
+
+    /// Returns the number of notes in the tree.
+    pub fn size(&self) -> usize {
+        self.parents.iter().enumerate().fold(
+            match (self.left, self.right) {
+                (None, None) => 0,
+                (Some(_), None) | (None, Some(_)) => 1,
```

Comment:
Is `(None, Some(_))` ever possible?

Reply:
Technically no; tree appending is lazy, so the state transition is:
`(None, None) -> (Some(_), None) <--> (Some(_), Some(_))`

I'll make it explicitly `unreachable!()`.